### ISSUE TITLE
PE2 Lvl4 

### ISSUE THREAD

**Issue by just-trying-plz** [nus-cs2030/2021-s1#367]
# Description 
Hello, I am just fumbling abit at level 4. I am coping with blurry eyes & tired soul now but I think I will get this in due time. Just putting here because I think a fresh pair of eyes can spot my mistakes faster. If any of you are able to help, that would be great!! ><

here is the qn prompt:
Level 4
Now add another action killTroll into actions.jsh. Look at the sample runs below for the behaviour of the action. Once again, the output (prepended with "-->") must be specified in actions.jsh and not in the respective class files. Also note that a killed troll will vanish from the room.

I am not that familiar with Streams but I try. Currently, I am filtering out the options when Troll is killed using Stream's filter(). I think it works but it is just the addition back into the original list when i used add(). In my add method, I take in an 'object' (Candle, Troll or Sword) which returns a new Room. 

Sorry if it sounds dumb but I think I just need time to map it out. If I am missing any details, let me know I can explain! 
## Topic: 
Specific Topics this question is related to (If Any) 
java immutability functional-programming PE2 idk

## Screenshots (if any): 
Insert Images here if necessary
![image](https://user-images.githubusercontent.com/71790053/99159797-08319700-271b-11eb-8f98-dcd240df70af.png)
I played around with the filter to check if the way I carried it out works. It works here because the final list (noTroll) does not contain Troll when my original Room had a Candle, Troll and Sword. 

![image](https://user-images.githubusercontent.com/71790053/99159827-547cd700-271b-11eb-9c98-02248ce62e69.png)
This is the wrong output I am dealing with. This is regarding the logic where there is a troll and i have a sword. In this part, I filter the troll and add the remaining entries to a new list which is to be returned as part of a new Room. 

I have debugged to print the noTroll list that items are supposed to go into. It is empty at some points. I suspect the way that i am adding the items back into the list. Am I doing it right? If not, what should I be doing instead?

It looks very jumbled up but i managed to pass the first 3 testX.jsh

Thank you..!

**Comment by tlylt** [comment]
Hey,
For a start, your level 4 test2 output is wrong since you haven't `takeSword` yet. (Probably something wrong with `killTroll`)
![image](https://user-images.githubusercontent.com/41845017/99160646-572ffa00-2724-11eb-9952-8e686d707751.png)

So you want to have a way to return a list of items in the room, but without the troll in it, to simulate killing the troll?
Initially I did something similiar, like this:
```
a.getStore().stream().filter(x -> !x instanceof Troll).collect(Collectors.toList());
```
>This is the wrong output I am dealing with. This is regarding the logic where there is a troll and i have a sword. In this part, I filter the troll and add the remaining entries to a new list which is to be returned as part of a new Room.

You are probably on the right track. For the cases, there are
1. NoTroll
2. Not Taken Sword
3. Troll and Taken Sword

From your screenshot, can tell that somehow `killTroll` is not adding objects back into the store and returned in a new Room...
maybe try to deal with the cases one by one... share more details about your `killTroll` if u can.

**Comment by wongzw** [comment]
You can also do this without using streams by just looping through the items in the room and removing the troll (and create a new room cause immutability). 

Maybe that would be easier for your implementation :)

**Comment by just-trying-plz** [comment]
hello! I found the problem!! yes, there was a initial logic problem with my killTroll! I misplaced one of my brackets in that line which changed the way it was interpreted! @tlylt  thank you! This helped me!

@wongzw you're right, I should have started with that to get a running implementation first! Thank you for helping! I was just hoping to reducing the number of lines of code since they may count by change in the number of lines.. : ( 

Thank you all <3
